README.md
shtml
=====

Common Lisp Server Side Includes (SSI). 

Please note that this project is not 100% SSI compatible.

I create this project to make it easier when edit html pages in SSI compatible html editor.

This project is based on http://lisp.ystok.ru/yhtml/html-template.html

CHANGELOG
Version 0.10.2
2011-12-29
Added TMPL_ELIF support.
Full copatibility is provided when *attributes-are-lisp-forms* is NIL.
Enhanced the test suit test.lisp.

Version 0.10.1
2009-05-19
For tags TMPL_ELSE and all ending tags /TMPL_...,
added optional text between the tag name and the closing marker "-->".
Within TMPL_EVAL, a keyword is treated as a constant, not as variable;
for keywords, no seeking via *value-access-function* in template-eval
any longer.

Version 0.10.0
2008-04-29
Added template-eval.lisp with additions and changes (by Dmitriy Ivanov)

Version 0.9.1
2007-11-16
Fixed bug in SKIP-LEADING-WHITESPACE (thanks to Chaitanya Gupta)
	
Version 0.9.0
2007-03-09
Hand down external format to included files (suggested by Igor Plekhov)

Version 0.8.0
2006-12-12
Added TMPL_CALL (thanks to Marijn Haverbeke)

Version 0.7.0
2006-09-30
Added TMPL_REPEAT

Version 0.6.0
2006-09-14
Changed default for *VALUE-ACCESS-FUNCTION*
Added *STRING-MODIFIER*	
Added ESCAPE-xxx functions (from CL-WHO)
Fixed docs for hyperdoc support
General documentation enhancements
Fixed test.lisp w.r.t. *FORMAT-NON-STRINGS*	
	
Version 0.5.0
2006-09-08
Added *FORMAT-NON-STRINGS*

Version 0.4.0
2006-04-03
Added TMPL_UNLESS (requested by Igor Plekhov)

Version 0.3.1
2005-08-05
Introduced *FORCE-DEFAULT*

Version 0.3.0
2005-07-03
Introduced *TEMPLATE-OUTPUT* so template printers don't interfer with *STANDARD-OUTPUT* (proposed by Norman Werner)
Small enhancements to docs

Version 0.2.0
2005-06-09
Added :DO-NOT-CACHE option to CREATE-TEMPLATE-PRINTER
Added clever DEFMETHOD trick by Peter Seibel
Fixed bug in READ-UNTIL which doesn't affect HTML-TEMPLATE
Better example in docs inspired by Bruce R. Lewis' critique on lisp-web
Mention asdf-install in docs
Added hyperdoc support
Added :HTML-TEMPLATE to *FEATURES*

Version 0.1.2
2003-07-15
Argh, typo in index.html... :(

Version 0.1.1
2003-07-15
Cleaner use of restarts thanks to James Anderson and Kent M. Pitman
Updated docs accordingly (removed stuff about OTHER-VALUE)
Removed ECL specific stuff (no longer necessary due to recent ECL fixes by Juan Jose Garcia Ripoll)
Corrected date of 0.1.0 release in CHANGELOG (don't laugh!)

Version 0.1.0
2003-07-15
Initial release


test.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: SHTML-TEST; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/shtml/test.lisp,v 1.13 2007/01/01 23:49:16 edi Exp $

;;; Copyright (c) 2003-2007, Dr. Edmund Weitz. All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package #:cl-user)

#-:cormanlisp
(defpackage #:shtml-test
  (:use #:cl #:shtml))

#+:cormanlisp
(defpackage "SHTML-TEST"
  (:use "CL" "SHTML"))

(in-package #:shtml-test)

(format t "~&Please wait a couple of seconds.")
(force-output)

(defvar tmp-dir #-(and lispworks win32) #p"/tmp/"
                #+(and lispworks win32) (pathname (lw:string-append
                                                   (or (lw:environment-variable "TEMP")
                                                       (lw:environment-variable "TMP")
                                                       "")
                                                   #\/)))

(defmacro failedp (&body body)
  `(handler-case
    (progn ,@body nil)
    (condition () t)))

(defmacro warnedp (&body body)
  `(handler-case
    (progn ,@body nil)
    (warning () t)))

;;; Restore Edi's favorites

(defun plist-value-access-function (symbol values &optional in-loop-p)
  (let ((result (getf values symbol)))
    (cond ((and in-loop-p *sequences-are-lists*)
           (loop for element in result
                 when (and element (listp element))
                 ;; keep values from upper levels
                 collect (append element values)
                 else
                         collect element))
          (t result))))

(setq *value-access-function* 'plist-value-access-function
      *template-package* *template-symbol-package*
      *attributes-are-lisp-forms* nil)


(defmacro test (result &rest args)
  (if result
      `(assert (string= ,result
                        (with-output-to-string (*default-template-output*)
                          (fill-and-print-template ,@args))))
      `(assert (failedp (fill-and-print-template ,@args)))))

(test "abc" "<!-- TMPL_VAR foo -->" '(:foo "abc"))
(test "abc" "<!-- tmpl_var foo -->" '(:foo "abc"))
(test "<!-- tmpl_vaar foo -->" "<!-- tmpl_vaar foo -->" '(:foo "abc"))
(test "xabcy" "x<!-- TMPL_VAR foo -->y" '(:foo "abc"))
(test "" "<!-- TMPL_VAR foo -->" nil)
(test "" "<!-- TMPL_VAR foo -->" '(foo "abc"))
(test "" "<!-- TMPL_VAR foo -->" '(:bar "abc"))
(test "abc" "<!-- TMPL_VAR 'foo' -->" '(:foo "abc"))
(test "abc" "<!-- TMPL_VAR \"foo\" -->" '(:foo "abc"))
(test nil "<!-- TMPL_VAR foo-->" '(:foo "abc"))
(test "" "<!-- TMPL_IF foo -->abc<!-- /TMPL_IF -->" nil)
(test "" "<!-- TMPL_IF foo -->abc<!-- /TMPL_IF -->" '(:foo nil))
(test "abc" "<!-- TMPL_IF foo -->abc<!-- /TMPL_IF -->" '(:foo t))
(test "abc" "<!-- TMPL_IF foo -->abc<!-- /TMPL_IF -->" '(:foo t :bar 42))
(test "" "<!-- TMPL_IF foo -->abc<!-- /TMPL_IF foo -->" nil)	; DI: Accepted!
(test nil "<!-- TMPL_IF -->abc<!-- /TMPL_IF -->" nil)
(test "def" "<!-- TMPL_IF foo -->abc<!-- TMPL_ELSE-->def<!-- /TMPL_IF -->" nil)
(test nil "<!-- TMPL_IF foo -->abc<!-- TMPL_ELSE-->def<!-- /TMPL_UNLESS -->" nil)
(test "def" "<!-- TMPL_IF foo -->abc<!-- TMPL_ELSE-->def<!-- /TMPL_IF -->" '(:foo nil))
(test "abc" "<!-- TMPL_IF foo -->abc<!-- TMPL_ELSE-->def<!-- /TMPL_IF -->" '(:foo t))
(test "abc" "<!-- TMPL_UNLESS foo -->abc<!-- TMPL_ELSE-->def<!-- /TMPL_UNLESS -->" '(:foo nil))
(test "def" "<!-- TMPL_UNLESS foo -->abc<!-- TMPL_ELSE-->def<!-- /TMPL_UNLESS -->" '(:foo t))
(test nil "<!-- TMPL_UNLESS foo -->abc<!-- TMPL_ELSE-->def<!-- /TMPL_IF -->" '(:foo t))
(test "abc" "<!-- TMPL_IF foo --><!-- TMPL_VAR foo --><!-- TMPL_ELSE-->def<!-- /TMPL_IF -->" '(:foo "abc"))
(test "def" "<!-- TMPL_IF foo --><!-- TMPL_VAR foo --><!-- TMPL_ELSE-->def<!-- /TMPL_IF -->" '(:foo nil))
(test "abcabcabc" "<!-- TMPL_IF foo --><!-- TMPL_VAR foo -->abc<!-- TMPL_VAR foo --><!-- TMPL_ELSE-->def<!-- /TMPL_IF -->" '(:foo "abc"))
(test "defdefdef" "<!-- TMPL_IF foo --><!-- TMPL_VAR foo -->abc<!-- TMPL_VAR foo --><!-- TMPL_ELSE--><!-- TMPL_VAR bar -->def<!-- TMPL_VAR bar --><!-- /TMPL_IF -->" '(:bar "def"))
(test "[]" "[<!-- TMPL_LOOP foo -->[x]<!-- /TMPL_LOOP -->]" '(:foo nil))
(test "[xxx]" "[<!-- TMPL_REPEAT foo -->x<!-- /TMPL_REPEAT -->]" '(:foo 3))
(test "[]" "[<!-- TMPL_REPEAT foo -->x<!-- /TMPL_REPEAT -->]" '(:foo 0))
(test "[]" "[<!-- TMPL_REPEAT foo -->x<!-- /TMPL_REPEAT -->]" '(:foo "foo"))
(test nil "[<!-- TMPL_REPEAT foo -->x<!-- /TMPL_LOOP -->]" '(:foo 3))
(test nil "[<!-- TMPL_LOOP foo -->x<!-- /TMPL_REPEAT -->]" '(:foo 3))
(test "[[x][x][x]]" "[<!-- TMPL_LOOP foo -->[x]<!-- /TMPL_LOOP -->]" '(:foo (1 2 3)))
(test "[[1][2][3]]" "[<!-- TMPL_LOOP foo -->[<!-- TMPL_VAR bar -->]<!-- /TMPL_LOOP -->]" '(:foo ((:bar "1") (:bar "2") (:bar "3"))))
(test "[[][][]]" "[<!-- TMPL_LOOP foo -->[<!-- TMPL_VAR bar -->]<!-- /TMPL_LOOP -->]" '(:foo (() () ())))
(test "[[1][2][3]]" "[<!-- TMPL_LOOP foo -->[<!-- TMPL_VAR bar -->]<!-- /TMPL_LOOP -->]" '(:foo ((:bar "1") (:bar "2") (:bar "3"))))
(test "[[1][][3]]" "[<!-- TMPL_LOOP foo -->[<!-- TMPL_VAR bar -->]<!-- /TMPL_LOOP -->]" '(:foo ((:bar "1") () (:bar "3"))))
(test "[[1][2][3]]" "[<!-- TMPL_LOOP foo -->[<!-- TMPL_IF 'bar' --><!-- TMPL_VAR bar --><!-- TMPL_ELSE-->2<!-- /TMPL_IF -->]<!-- /TMPL_LOOP -->]" '(:foo ((:bar "1") () (:bar "3"))))
(test "[[123][456][789]]" "[<!-- TMPL_LOOP 'foo' -->[<!-- TMPL_LOOP 'bar' --><!-- TMPL_VAR 'bar' --><!-- /TMPL_LOOP -->]<!-- /TMPL_LOOP -->]" '(:foo ((:bar ((:bar "1") (:bar "2") (:bar "3")))
                                                                                                                                                      (:bar ((:bar "4") (:bar "5") (:bar "6")))
                                                                                                                                                      (:bar ((:bar "7") (:bar "8") (:bar "9"))))))
(test "[[123][baz][789]]" "[<!-- TMPL_LOOP 'foo' -->[<!-- TMPL_IF baz --><!-- TMPL_LOOP 'baz' --><!-- TMPL_VAR 'bar' --><!-- /TMPL_LOOP --><!-- TMPL_ELSE -->baz<!-- /TMPL_IF -->]<!-- /TMPL_LOOP -->]" '(:foo ((:baz ((:bar "1") (:bar "2") (:bar "3")))
                                                                                                                                                      ()
                                                                                                                                                      (:baz ((:bar "7") (:bar "8") (:bar "9"))))))
(test nil "<!-- TMPL_ELSE -->" nil)
(test "<!-- /TMPL_ELSE -->" "<!-- /TMPL_ELSE -->" nil)
(test nil "<!-- /TMPL_IF -->" nil)
(test nil "<!-- /TMPL_UNLESS -->" nil)
(test nil "<!-- /TMPL_LOOP -->" nil)
(test nil "<!-- TMPL_IF foo --><!-- TMPL_ELSE -->" nil)
(test nil "<!-- TMPL_UNLESS foo --><!-- TMPL_ELSE -->" nil)
(test nil "<!-- TMPL_LOOP foo --><!-- TMPL_ELSE -->" nil)
(test nil "<!-- TMPL_LOOP foo --><!-- TMPL_ELSE --><!-- /TMPL_LOOP -->" nil)
(test nil "<!-- TMPL_IF bar --><!-- TMPL_LOOP foo --><!-- TMPL_ELSE --><!-- /TMPL_LOOP -->" nil)
(test nil "<!-- TMPL_IF foo --><!-- TMPL_IF bar -->1<!-- TMPL_ELSE -->2<!-- /TMPL_IF --><!-- TMPL_ELSE --><!-- TMPL_IF baz -->3<!-- TMPL_ELSE -->4<!-- /TMPL_IF -->" nil)
(test "1" "<!-- TMPL_IF foo --><!-- TMPL_IF bar -->1<!-- TMPL_ELSE -->2<!-- /TMPL_IF --><!-- TMPL_ELSE --><!-- TMPL_IF baz -->3<!-- TMPL_ELSE -->4<!-- /TMPL_IF --><!-- /TMPL_IF -->" '(:foo t :bar t))
(test "2" "<!-- TMPL_IF foo --><!-- TMPL_IF bar -->1<!-- TMPL_ELSE -->2<!-- /TMPL_IF --><!-- TMPL_ELSE --><!-- TMPL_IF baz -->3<!-- TMPL_ELSE -->4<!-- /TMPL_IF --><!-- /TMPL_IF -->" '(:foo t :bar nil))
(test "3" "<!-- TMPL_IF foo --><!-- TMPL_IF bar -->1<!-- TMPL_ELSE -->2<!-- /TMPL_IF --><!-- TMPL_ELSE --><!-- TMPL_IF baz -->3<!-- TMPL_ELSE -->4<!-- /TMPL_IF --><!-- /TMPL_IF -->" '(:foo nil :baz t))
(test "4" "<!-- TMPL_IF foo --><!-- TMPL_IF bar -->1<!-- TMPL_ELSE -->2<!-- /TMPL_IF --><!-- TMPL_ELSE --><!-- TMPL_IF baz -->3<!-- TMPL_ELSE -->4<!-- /TMPL_IF --><!-- /TMPL_IF -->" '(:foo nil :baz nil))
(test "X" "<!-- TMPL_CALL foo -->" '(:foo (("X"))))
(test "QUUX" "<!-- TMPL_VAR baz --><!-- TMPL_CALL foo -->" '(:baz "Q"
                                                             :foo (("<!-- TMPL_VAR bar -->" :bar "U")
                                                                   ("<!-- TMPL_VAR bar -->X" :bar "U"))))
(test "" "<!-- TMPL_IF foo --><!-- TMPL_CALL bar --><!-- /TMPL_IF -->" '(:foo (("---"))))
(test nil "<!-- TMPL_CALL foo -->" '(:foo 57))

(let ((temp-name (make-pathname :name (format nil "template-test-~A" (random 1000000))
                                :defaults tmp-dir)))
  (with-open-file (stream temp-name :direction :output :if-exists :error)
    (write-string "<!-- TMPL_VAR foo -->" stream))
  (let ((*warn-on-creation* nil))
    (test "abc" temp-name '(:foo "abc")))
  (with-open-file (stream temp-name :direction :input)
    (test "def" stream '(:foo "def")))
  (with-open-file (stream temp-name :direction :input)
    (let ((tp (create-template-printer stream)))
      (test "ghi" tp '(:foo "ghi"))))
  (let ((tp (create-template-printer temp-name)))
    (test "jkl" tp '(:foo "jkl")))
  (let ((tp (create-template-printer "<!-- TMPL_VAR foo -->")))
    (test "mno" tp '(:foo "mno")))
  (delete-file temp-name)
  ;; sleep because of FILE-WRITE-DATE
  (sleep 2)
  (with-open-file (stream temp-name :direction :output :if-exists :error)
    (write-string "<!-- TMPL_VAR bar -->" stream))
  (assert (warnedp (create-template-printer temp-name)))
  (assert (not (warnedp (create-template-printer temp-name))))
  (assert (warnedp (create-template-printer temp-name :force t)))
  (delete-from-template-cache temp-name)
  (assert (warnedp (create-template-printer temp-name)))
  (clear-template-cache)
  (assert (warnedp (create-template-printer temp-name)))
  (delete-file temp-name))

(let ((*template-start-marker* "<")
      (*template-end-marker* ">"))
  (test "The quick <brown> fox" "The <TMPL_VAR 'speed'> <brown> fox"
        '(:speed "quick")))

(let* ((random-string (format nil "template-test-~A" (random 1000000)))
       (temp-name (merge-pathnames random-string tmp-dir))
       (*default-template-pathname* tmp-dir))
  (with-open-file (stream temp-name :direction :output :if-exists :error)
    (write-string "The <!-- TMPL_VAR speed --> brown fox" stream))
  (let ((*warn-on-creation* nil))
    (test "The very fast brown fox"
          (make-pathname :name random-string)
          '(:speed "very fast")))
  (delete-file temp-name)
  ;; sleep because of FILE-WRITE-DATE
  (sleep 2)
  (with-open-file (stream temp-name :direction :output :if-exists :error)
    (write-string "The <!-- TMPL_VAR speed --> brown fox" stream))
  (let ((*warn-on-creation* nil))
    (test "The very fast brown fox"
          (format nil "<!-- TMPL_INCLUDE '~A' -->" random-string)
          '(:speed "very fast")))
  (delete-file temp-name))

(let* ((random-string (format nil "template-test-~A" (random 1000000)))
       (temp-name (merge-pathnames random-string tmp-dir))
       (random-string-2 (format nil "template-test-2-~A" (random 1000000)))
       (temp-name-2 (merge-pathnames random-string-2 tmp-dir))
       (*default-template-pathname* tmp-dir))
  (with-open-file (stream temp-name :direction :output :if-exists :error)
    (format stream "<!-- TMPL_INCLUDE '~A' -->" random-string-2))
  (with-open-file (stream temp-name-2 :direction :output :if-exists :error)
    (format stream "<!-- TMPL_INCLUDE '~A' -->" random-string))
  (test nil (format nil "<!-- TMPL_INCLUDE '~A' -->" random-string) nil)
  (delete-file temp-name)
  (delete-file temp-name-2))

(assert (string= "The slow brown fox"
                 (with-output-to-string (stream)
                   (let ((*default-template-output* stream))
                     (fill-and-print-template "The <!-- TMPL_VAR speed --> brown fox"
                                              '(:speed "slow"))))))

(let* ((tp (create-template-printer "The <!-- TMPL_VAR speed --> brown fox"))
       (*convert-nil-to-empty-string* nil))
  (with-output-to-string (*default-template-output*)
    (test nil tp '(:foo "bar"))))

(let ((tp (create-template-printer "The <!-- TMPL_VAR speed --> brown fox")))
  (handler-bind
    ((template-missing-value-error (lambda (condition)
                                     (declare (ignore condition))
                                     (use-value "slow"))))
    (let ((*convert-nil-to-empty-string* nil))
      (test "The slow brown fox" tp '(:foo "bar")))))

(let ((*sequences-are-lists* nil))
  (test "[1][2][3]"
        "<!-- TMPL_LOOP vector -->[<!-- TMPL_VAR item -->]<!-- /TMPL_LOOP -->"
        '(:vector #((:item "1")
                    (:item "2")
                    (:item "3"))))
  (test "QUUX" "<!-- TMPL_VAR baz --><!-- TMPL_CALL foo -->"
        '(:baz "Q"
          :foo #(("<!-- TMPL_VAR bar -->" :bar "U")
                 ("<!-- TMPL_VAR bar -->X" :bar "U")))))

(let ((*upcase-attribute-strings* nil))
  (test "The slow brown fox"
        "The <!-- TMPL_VAR speed --> brown fox"
        '(:speed "quick" :|speed| "slow")))

(let ((*template-symbol-package* *package*))
  (test "The slow brown fox"
        "The <!-- TMPL_VAR speed --> brown fox"
        '(:speed "quick" speed "slow")))

(let ((tp (create-template-printer "The <!-- TMPL_VAR speed --> brown fox"))
      (*value-access-function* #'gethash)
      (hash (make-hash-table :test #'eq)))
  (setf (gethash :speed hash) "fast")
  (test "The fast brown fox" tp hash))

(let ((values (list :row-loop
                    (loop for row in '((1 2 3 4) (2 3 4 5) (3 4 5 6))
                          collect (list :col-loop
                                        (loop for col in row
                                              collect (list :item
                                                            (format nil "~A" col)))))))
      (template "<table>
  <!-- TMPL_LOOP row-loop -->
  <tr>
    <!-- TMPL_LOOP col-loop -->
    <td><!-- TMPL_VAR item --></td>
    <!-- /TMPL_LOOP -->
  </tr>
  <!-- /TMPL_LOOP -->
</table>")
      (result "<table>
  <tr>
    <td>1</td>
    <td>2</td>
    <td>3</td>
    <td>4</td>
  </tr>
  <tr>
    <td>2</td>
    <td>3</td>
    <td>4</td>
    <td>5</td>
  </tr>
  <tr>
    <td>3</td>
    <td>4</td>
    <td>5</td>
    <td>6</td>
  </tr>
</table>")
      (*ignore-empty-lines* t))
  (test result template values))

(let ((tp (create-template-printer "A square has <!-- TMPL_VAR number --> corners"))
      (*format-non-strings* nil))
  (handler-bind
    ((template-not-a-string-error (lambda (condition)
                                    (use-value
                                     (format nil "~R"
                                             (template-not-a-string-error-value condition))))))
    (test "A square has four corners" tp '(:number 4))))

;;; TMPL_ELIF

(test nil "<!-- TMPL_UNLESS foo -->xfoo<!-- TMPL_ELIF bar -->xbar<!-- /TMPL_UNLESS -->" ())
(test nil "<!-- TMPL_IF foo -->xfoo<!-- TMPL_ELIF bar -->xbar<!-- /TMPL_UNLESS -->" ())

(test "xfooy" "x<!-- TMPL_IF foo -->foo<!-- TMPL_ELIF bar -->bar<!-- /TMPL_IF -->y"
  '(:foo t :bar nil))
(test "xy" "x<!-- TMPL_IF foo -->foo<!-- TMPL_ELIF bar -->bar<!-- /TMPL_IF -->y"
  ())
(test "xbary" "x<!-- TMPL_IF foo -->foo<!-- TMPL_ELIF bar -->bar<!-- /TMPL_IF -->y"
  '(:foo nil :bar t))
(test "xbazy" "x<!-- TMPL_IF foo -->foo<!-- TMPL_ELIF bar -->bar<!-- TMPL_ELSE baz -->baz<!-- /TMPL_IF -->y"
  ())

(test "xy" "x<!-- TMPL_IF foo -->foo<!-- TMPL_ELIF bar -->bar<!-- TMPL_IF baz -->baz<!-- /TMPL_IF --><!-- /TMPL_IF -->y"
  ())
(test "xbary" "x<!-- TMPL_IF foo -->foo<!-- TMPL_ELIF bar -->bar<!-- TMPL_IF baz -->baz<!-- /TMPL_IF --><!-- /TMPL_IF -->y"
  '(:bar t))
(test "xbarbazy" "x<!-- TMPL_IF foo -->foo<!-- TMPL_ELIF bar -->bar<!-- TMPL_IF baz -->baz<!-- /TMPL_IF --><!-- /TMPL_IF -->y"
  '(:bar t :baz t))


(format t "~&All tests passed...")


load.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: CL-USER; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/shtml/load.lisp,v 1.7 2007/01/01 23:49:16 edi Exp $

;;; Copyright (c) 2003-2007, Dr. Edmund Weitz.  All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package #:cl-user)

(let ((shtml-base-directory
        (make-pathname :name nil :type nil :version nil
                       :defaults (parse-namestring *load-truename*)))
      must-compile)
  (with-compilation-unit ()
    (loop for file in '("packages"
                        "specials"
                        "errors"
                        "util"
                        "template"
                        "api")
          do (let ((pathname (make-pathname :name file :type "lisp" :version nil
                                            :defaults shtml-base-directory)))
               #-:cormanlisp
               (let ((compiled-pathname (compile-file-pathname pathname)))
                 (unless (and (not must-compile)
                              (probe-file compiled-pathname)
                              (< (file-write-date pathname)
                                 (file-write-date compiled-pathname)))
                   (setq must-compile t)
                   (compile-file pathname))
                 (setq pathname compiled-pathname))
               (load pathname)))))


specials.lisp
;;; -*- Mode: LISP; Syntax: COMMON-LISP; Package: SHTML; Base: 10 -*-
;;; $Header: /usr/local/cvsrep/shtml/specials.lisp,v 1.24 2007/03/09 13:09:16 edi Exp $

;;; Copyright (c) 2003-2007, Dr. Edmund Weitz. All rights reserved.

;;; Redistribution and use in source and binary forms, with or without
;;; modification, are permitted provided that the following conditions
;;; are met:

;;;   * Redistributions of source code must retain the above copyright
;;;     notice, this list of conditions and the following disclaimer.

;;;   * Redistributions in binary form must reproduce the above
;;;     copyright notice, this list of conditions and the following
;;;     disclaimer in the documentation and/or other materials
;;;     provided with the distribution.

;;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR 'AS IS' AND ANY EXPRESSED
;;; OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
;;; WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
;;; DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
;;; GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
;;; WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
;;; NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
;;; SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

(in-package #:shtml)

(defvar *find-string-hash* (make-hash-table :test #'equal)
  "Hash tables used internally by READ-UNTIL to cache offset arrays.")

(defvar *template-start-marker* "<!--"
  "The string template tags must start with")
(defvar *template-end-marker* "-->"
  "The string template tags must end with")

(defvar *printer-hash* (make-hash-table :test #'equal)
  "The cache for template printers.
Each entry is of the form (PRINTER . WRITE-DATE).")

(defvar *default-template-pathname* (make-pathname)
  "Each pathname is merged with this value before it is used by
CREATE-TEMPLATE-PRINTER.")

(defvar *default-template-output* *standard-output*
  "The output stream used by FILL-AND-PRINT-TEMPLATE when no STREAM
keyword was provided.")

(defvar *template-output* nil
  "The output stream that's used internally.")

(defvar *convert-nil-to-empty-string* t
  "Controls whether NIL values should resolve to empty strings or
raise an error.")

(defvar *format-non-strings* t
  "Controls whether TMPL_VAR will accept values which aren't
strings and convert them using \(FORMAT NIL \"~A\" ...).")

(defvar *sequences-are-lists* t
  "Controls whether TMPL_LOOP printers expect lists or vectors.")

(defvar *upcase-attribute-strings* t
  "Controls whether attribute strings associated with template tags
are upcased before they are interned.")

(defvar *no-cache-check* nil
  "Controls whether the FILE-WRITE-DATE check will be circumvented
when using FILL-AND-PRINT-TEMPLATE.")

(defvar *template-dependence* (make-hash-table :test #'equal)
  "cache templace depend on which templates")

(defvar *template-symbol-package* (find-package '#:keyword)
  "The package symbols are interned into.")

(defvar *ignore-empty-lines* nil
  "Controls whether template tags on their own lines produce empty
lines or not.")

(defvar *warn-on-creation* t
  "Controls whether a warning should be signaled if a new template
printer is created from a pathname argument.")

(defvar *current-line* 1
  "Internal line counter of the parser.")
(defvar *current-column* 0
  "Internal column counter of the parser.")

(defvar *included-files* nil
  "Internally used by CREATE-TEMPLATE-PRINTER-AUX to avoid infinite
TMPL_INCLUDE loops.")

(defvar *current-creating-template-path* nil
  "the creating template path")

(defvar *external-format* :default
  "The external format used when opening files.")

(defvar *uglify-js-p* t)

(defvar *value-access-function*
  (lambda (symbol environment &optional in-loop-p)
    (let ((result #+ylib (yl:cdr-assoq symbol environment)	; getf has originally used
                  #-ylib (cdr (assoc symbol environment))))
      (if (and in-loop-p shtml:*sequences-are-lists*)
          (loop for element in result			; keep values from upper levels
                collect (if (consp element) (append element environment) element))
          result)))
  "The function which associates \(attribute) symbols with their values.")

#|;; Sorry about replacing Edi's preference
(defvar *value-access-function*
  (lambda (symbol values &optional in-loop-p)
    (let ((result (getf values symbol)))
      (cond ((and in-loop-p *sequences-are-lists*)
             (loop for element in result
                   when (and element (listp element))
                     ;; keep values from upper levels
                     collect (append element values)
                   else
                     collect element))
            (t result))))
  "The function which associates \(attribute) symbols with their
values.")
|#

(defvar *call-template-access-function* #'car
  "Accessor function for extracting the called template from a
TMPL_CALL form.")

(defvar *call-value-access-function* #'cdr
  "Accessor function for extracting the values from a TMPL_CALL
form.")

(defvar *force-default* nil
  "The default value for the FORCE keyword argument to
CREATE-TEMPLATE-PRINTER.")

(defvar *string-modifier* 'escape-string-iso-8859-1
  "The function which is applied to strings which replace
TMPL_VAR tags.  Use #'CL:IDENTITY if you don't want to change the
strings.")

(defparameter *escape-char-p*
  #'(lambda (char)
      (or (find char "<>&'\"")
          (> (char-code char) 127)))
  "Used by ESCAPE-STRING to test whether a character should be escaped.")


;;; TEMPLATE-EVAL related stuff

(defvar *template-package* (find-package :cl-user)
  "The package symbols are interned into when reading forms
from template or HTML source files.")

(defvar *template-environment* ()
  "Stores global template variable bindings and is used by
template-eval as a default parameter value.")

;; Compatibility parameter used by create-template-printer-aux.
;; Its value does not matter for the below tags:
;; - TMPL_EVAL always treats "tag-rest" as a form,
;; - TMPL_VAR  always treats "tag-rest" as a plain attribute.

(defparameter *attributes-are-lisp-forms* t
  "When true, attributes are read and interpreted as Lisp forms for most
of the following tags:
  TMPL_IF, TMPL_ELIF, TMPL_UNLESS, TMPL_LOOP, TMPL_REPEAT, and TMPL_CALL.
When false, the reader works on plane attribute syntax only:
accepts symbols and interns them into *TEMPLATE-SYMBOL-PACKAGE*;
filling templates also proceeds in old fashion.")


;; stuff for Nikodemus Siivola's HYPERDOC
;; see <http://common-lisp.net/project/hyperdoc/>
;; and <http://www.cliki.net/hyperdoc>
;; also used by LW-ADD-ONS

(defvar *hyperdoc-base-uri* "http://weitz.de/shtml/")

(let ((exported-symbols-alist
       (loop for symbol being the external-symbols of :shtml
             collect (cons symbol
                           (concatenate 'string
                                        "#"
                                        (string-downcase symbol))))))
  (defun hyperdoc-lookup (symbol type)
    (declare (ignore type))
    (cdr (assoc symbol
                exported-symbols-alist
                :test #'eq))))
